server/hardware.c

```diff
@@ -579,23 +595,47 @@ const uint32_t *hw_read_table_data(struct hw_table *table, unsigned int number)
 }
 
 
-void hw_write_table(
+bool hw_table_supports_streaming(struct hw_table *table)
+{
+    return table->table_type == LONG_TABLE;
+}
+
+
+error__t hw_reset_table(struct hw_table *table, unsigned int number)
+{
+    if (table->table_type == LONG_TABLE)
+        return hw_long_table_write(
+            table->long_table.block_ids[number], NULL, 0, false, false);
+    else
+        return ERROR_OK;
+}
+
+
+error__t hw_write_table(
     struct hw_table *table, unsigned int number,
-    size_t offset, const uint32_t data[], size_t length)
+    const uint32_t data[], size_t length, bool streaming_mode, bool last_table)
 {
     switch (table->table_type)
     {
         case SHORT_TABLE:
-            memcpy(table->data[number] + offset,
-                data, length * sizeof(uint32_t));
-            write_short_table(table, number, offset + length);
+            memcpy(table->data[number], data, length * sizeof(uint32_t));
+            return write_short_table(table, number, length);
             break;
         case LONG_TABLE:
-            hw_long_table_write(
-                table->long_table.block_ids[number], data,
-                length * sizeof(uint32_t), offset * sizeof(uint32_t));
+            /* If length is zero, the table was already reset earlier */
+            if (length > 0) {
+                if (!streaming_mode)
+                    /* Only in fixed mode we should expose the data */
+                    memcpy(
+                        table->data[number], data, length * sizeof(uint32_t));
```

Comment:
Can we be confident at this point that `length` is bounded by the size of `table->data`?
I'm guessing so as there is an unchecked `memcpy` above under `SHORT_TABLE` which is inherited from previous code, but a double check of the code path leading here would be wise!

Reply:
Yes, this is only called from `table.c:complete_table_write`